Improve the generation of config.cmake and version.cmake files

jvgomez opened the pull request:
This PR addresses a problem with the CMake configuration files: paths of the libraries were hardcoded so that crosscompilation was difficult. Now, using the CMake function `CONFIGURE_PACKAGE_CONFIG_FILE` the paths are relative. In my tests,it works in my PC while also being able to properly find the libraries when crosscompiling.
Additionally, I used the tool `WRITE_BASIC_PACKAGE_VERSION_FILE` to simplify the version.cmake file generation.

Issues I noticed (but I didn't have time to fix):
- CMakeLists.txt are not following any convention (mixing uppercase and lowercase).
- Dynamic-edt-3d name is confusing, sometimes using kind of camel-case, sometimes using underscores and some other times dashes.
- There is a problem with the Qt CMake variables in Octovis` CMakeLists.txt: `  "@QGLViewer_LIBRARIES@"
  "@QT_LIBRARIES@"` as they use platform specific libraries. But I guess it is not very important as it is "just" a visualization tool.
I can open an issue with these if you want. But just wanted to mention them here.

jvgomez commented:
I see that the CI fails because `CMake 2.8.8 or higher is required.  You are running version 2.8.7`. The two CMake functions I are introduced in 2.8.8. Is it a problem to update it to 2.8.8?

ahornung commented:
The CI build is based on Ubuntu Precise, where 2.8.7 is the supplied CMake version.

Could you try adding
```
sudo: required
dist: trusty
```
to `.travis.yml` in your PR as described [here](https://docs.travis-ci.com/user/trusty-ci-environment/)?

ahornung commented:
Thanks for the PR! If I understand it correctly then the new(ish) CMake functionality of `CONFIGURE_PACKAGE_CONFIG_FILE` allows to make the package relocatable without using relative paths (as in #104).

Testing locally with a newer CMake version, the build fails however. It appears that there is no more CMakeConfig and version config written for the build tree. There are four use cases that need work:

1. Compile and use the build tree of the distribution (from `build` inside the repo)
2. Compile and install the complete distribution (from `build` inside the repo)
3. Compile and use the build tree of a single package (e.g. from `octomap\build`)
4. Compile and install the build tree of a single package (e.g. from `octomap\build`)

1.) fails because octovis can't find the config for octomap.

The other points are best covered with issues of their own

ahornung commented on dynamicEDT3D/CMakeLists.txt:
```diff
@@ -96,10 +96,6 @@ ELSE()
-  "${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/cmake/dynamicEDT3D/dynamicEDT3DConfig.cmake" @ONLY)
-configure_file(dynamicEDT3DConfig-version.cmake.in
-  "${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/cmake/dynamicEDT3D/dynamicEDT3DConfig-version.cmake" @ONLY)
 
```
These files are missing from the build tree

jvgomez commented:
This is weird. Using CMake 2.8.12.2, I did the following:

```
$ git clone https://github.com/rapyuta/octomap.git octomap_rapyuta
$ cd octomap_rapyuta && git checkout config_fix 
$ mkdir build && cd build
$ cmake ..
```

I got the following warnings:
```
-- Configuring done
CMake Warning at octovis/CMakeLists_src.txt:100 (add_executable):
  Cannot generate a safe runtime search path for target octovis because there
  is a cycle in the constraint graph:

    dir 0 is [/home/jvgomez/Desktop/octomap_rapyuta/octovis/src/extern/QGLViewer]
    dir 1 is [/home/jvgomez/Desktop/octomap/lib]
      dir 2 must precede it due to runtime library [liboctovis.so.1.8]
    dir 2 is [/home/jvgomez/Desktop/octomap_rapyuta/lib]
      dir 1 must precede it due to runtime library [liboctomap.so.1.8]

  Some of these libraries may not be found correctly.
Call Stack (most recent call first):
  octovis/CMakeLists.txt:85 (INCLUDE)


CMake Warning at dynamicEDT3D/src/examples/CMakeLists.txt:1 (add_executable):
  Cannot generate a safe runtime search path for target exampleEDT3D because
  there is a cycle in the constraint graph:

    dir 0 is [/home/jvgomez/Desktop/octomap_rapyuta/lib]
      dir 1 must precede it due to runtime library [liboctomap.so.1.8]
    dir 1 is [/home/jvgomez/Desktop/octomap/lib]
      dir 0 must precede it due to runtime library [libdynamicedt3d.so.1.8]

  Some of these libraries may not be found correctly.


CMake Warning at dynamicEDT3D/src/examples/CMakeLists.txt:4 (add_executable):
  Cannot generate a safe runtime search path for target exampleEDTOctomap
  because there is a cycle in the constraint graph:

    dir 0 is [/home/jvgomez/Desktop/octomap_rapyuta/lib]
      dir 1 must precede it due to runtime library [liboctomap.so.1.8]
    dir 1 is [/home/jvgomez/Desktop/octomap/lib]
      dir 0 must precede it due to runtime library [libdynamicedt3d.so.1.8]

  Some of these libraries may not be found correctly.
```

But still, after `$ make -j8` everything built OK and I can see .cmake files at `octomap_rapyuta/build/dynamicEDT3D/InstallFiles` (and for Octovis and Octomap as well).

jvgomez commented:
Update:
I am actually able to reproduce the problem. It was working because I had another clone of octomap already compiled in the same folder and also the ROS-octomap package. Debugging.

jvgomez commented:
Fixed! As I said I am not an expert with CMake so I just removed some lines because I thought they were duplicated :) But now I updated them with the new syntax as well.

ahornung commented:
Thanks for the effort, works here as well!
